File: 0x00 基本算法/0x01 位运算/AcWing998.cpp
#include <bits/stdc++.h>
using namespace std;
typedef pair<string, int> PSI;
const int N = 1e5 + 10;
PSI p[N];
int n, m;

int calc(int bit, int now) {
  for (int i = 0; i < n; i++) {
    auto [op, x] = p[i];
    int b = x >> bit & 1;
    if (op == "AND") now &= b;
    else if (op == "OR") now |= b;
    else now ^= b;
  }
  return now;
}

int main() {
  cin >> n >> m;
  for (int i = 0; i < n; i ++ ) cin >> p[i].first >> p[i].second;

  int t = 0, res = 0; // 每位独立，贪心让伤害最大（每位上都倾向于选0，这样初始值才更有可能在m范围内）
  for (int i = 30; i >= 0; i -- ) { // 本题中m最大是10^9，lg2(10^9) = 3log2(10^3) < 30
    int a = calc(i, 0), b = calc(i, 1);
    if (t + (1 << i) <= m && a < b) t += 1 << i, res += b << i;
    else res += a << i;
  }
  cout << res << endl;
  return 0;
}

File: 0x00 基本算法/0x01 位运算/AcWing998.py
import sys

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
        
    n = int(input_data[0])
    m = int(input_data[1])
    
    p = [(input_data[i], int(input_data[i + 1])) for i in range(2, 2 + n * 2, 2)]
        
    def calc(bit, now):
        for op, x in p:
            b = (x >> bit) & 1
            if op == "AND":
                now &= b
            elif op == "OR":
                now |= b
            else:
                now ^= b
        return now

    t = res = 0
    for i in range(30, -1, -1):
        a, b = calc(i, 0), calc(i, 1)
        if t + (1 << i) <= m and a < b:
            t += 1 << i
            res += b << i
        else:
            res += a << i
            
    print(res)

if __name__ == '__main__':
    main()